fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBIndexOutput.java

```diff
@@ -0,0 +1,150 @@
+
+    static {
+        BUFFERS = new ArrayBlockingQueue<>(POOL_SIZE);
+        for (int i = 0; i < POOL_SIZE; i++) {
+            BUFFERS.add(ByteBuffer.allocate(FDBDirectory.DEFAULT_BLOCK_SIZE));
+        }
+    }
+
+    /**
+     * Create an FDBIndexOutput given a resource description, name, and FDBDirectory.
+     *
+     * @param resourceDescription description of resource
+     * @param name name of resource
+     * @param fdbDirectory existing FDBDirectory
+     */
+    public FDBIndexOutput(String resourceDescription, String name, FDBDirectory fdbDirectory) {
+        super(resourceDescription, name);
+        LOG.trace("init() -> resource={}, name={}", resourceDescription, name);
+        this.resourceDescription = resourceDescription;
+        this.fdbDirectory = fdbDirectory;
+        blockSize = fdbDirectory.getBlockSize();
+        buffer = BUFFERS.poll();
+        if (buffer == null) {
+            buffer = ByteBuffer.allocate((int)blockSize);
+        }
+        crc = new CRC32();
+        id = fdbDirectory.getIncrement();
+    }
+
+    /**
+     * Close the directory which writes the FileReference.
+     */
+    @Override
+    @SpotBugsSuppressWarnings(value = "RV_RETURN_VALUE_IGNORED_BAD_PRACTICE", justification = "it is fine if it is not accepted")
+    public void close() {
+        LOG.trace("close() -> resource={}", resourceDescription);
+        flush();
+        fdbDirectory.writeFDBLuceneFileReference(resourceDescription, new FDBLuceneFileReference(id, currentSize, blockSize));
+        BUFFERS.offer(buffer);
+    }
+
+    @Override
+    public long getFilePointer() {
+        LOG.trace("getFilePointer() -> resource={}, pointer={}", resourceDescription, currentSize);
+        return currentSize;
+    }
+
+    @Override
+    public long getChecksum() {
+        LOG.trace("getChecksum() -> resource={}, checksum={}", resourceDescription, crc.getValue());
+        return crc.getValue();
+    }
+
+    @Override
+    public void writeByte(final byte b) {
+        LOG.trace("writeByte() -> resource={}", resourceDescription);
+        buffer.put(b);
+        crc.update(b);
+        currentSize++;
+        if (currentSize % blockSize == 0) {
+            flush();
+        }
+    }
+
+    @Override
+    public void copyBytes(final DataInput input, final long numBytes) throws IOException {
+        LOG.debug("copy bytes input={}, numbytes={}", input, numBytes);
+        super.copyBytes(input, numBytes);
+    }
+
+    @Override
+    public void writeBytes(final byte[] bytes, final int offset, final int length) {
+        LOG.trace("writeBytes() -> resource={}, offset={}, length={}", resourceDescription, offset, length);
+        crc.update(bytes, offset, length);
+        int bytesWritten = 0;
+        while (bytesWritten < length) {
+            int toWrite = (int) (length - bytesWritten + (currentSize % blockSize) > blockSize ? blockSize - (currentSize % blockSize) : length - bytesWritten);
```

Comment:
not 100% certain, but think this could probably be simplified to just refer to `bytesWritten`. If not, perhaps a quick note about what the purpose of `currentSize` is?

Reply:
Added notes on it.  Added a note about writeBytes being called many times.